File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);

        int pre = 1, post = 1;

        // Pass 1: prefix products
        for (int i = 0; i < n; i++) {
            res[i] = pre;
            pre = pre * nums[i];
        }

        // Pass 2: multiply suffix products
        for (int i = n - 1; i >= 0; i--) {
            res[i] = post * res[i];
            post = post * nums[i];
        }

        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        # Initialize the result array with 1s of the same length as nums
        res = [1] * (len(nums))
        
        # Pass 1: Calculate left products for each element
        # 'prefix' keeps track of the product of all elements to the left
        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
            
        # Pass 2: Calculate right products and multiply with left products in place
        # 'suffix' keeps track of the product of all elements to the right
        suffix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= suffix
            suffix *= nums[i]
            
        # Return the final product array where each element is the product of all other elements
        return res